### ISSUE TITLE
[patterns] castPatterns take a `type` not a `typeNotVoid`

### ISSUE THREAD

**Issue by modulovalue** [dart-lang/language#2897]
I've noticed that castPatterns accept void on their rhs. However, the rhs of typeCasts does not i.e.:

<img width="1361" alt="Bildschirm­foto 2023-03-10 um 14 13 06" src="https://user-images.githubusercontent.com/24444584/224325680-1694314e-b65c-40b5-8be5-6f79bbab3481.png">

https://github.com/dart-lang/sdk/blob/4751ed0bb8675402c66d530b99fff94e327e175c/tools/spec_parser/Dart.g#L1073-L1075

https://github.com/dart-lang/sdk/blob/4751ed0bb8675402c66d530b99fff94e327e175c/tools/spec_parser/Dart.g#L1031-L1033

Here's the relevant section in the feature specification: [cast pattern](https://github.com/dart-lang/language/blob/master/accepted/future-releases/0546-patterns/feature-specification.md#cast-pattern)

I'm wondering, wouldn't it make sense to make castPatterns also not take `void` on their rhs or are there use cases where it would make sense to make castPatterns support casting to void? (_of course, a typedef can be used to cast to void, this is more about consistency to prevent questions like 'why can I do X but not Y?'_)

**Comment by eernstg** [comment]
Exactly, a type alias like `typedef Void = void;` allows the type `void` to occur in a type test like `e is Void` or type checks like `e as Void`.

The only reason why it isn't supported is history: `void` was _only_ used as a return type (way back in 2014 or so), and `void` is a reserved word, so it wasn't actually a `<type>`. There was no reason to do anything special in order to allow `e is void` at te time. But today we _can_ do that, semantically, and there is no point in trying to prohibit it.

I think this means that it's no problem that the `as` pattern allows `void`. We could then generalize the `is` and `as` expressions such that they will also allow `void`, syntactically, if it turns out to be useful. That is, if someone says "why can I do X but not Y?" at some point in the future, and also gives a good use case.

**Comment by modulovalue** [comment]
> There was no reason to do anything special in order to allow e is void at the time. But today we can do that, semantically, and there is no point in trying to prohibit it.

I see, thank you for providing the historical context.

> I think this means that it's no problem that the as pattern allows void

I agree, because I would prefer for `typeTest`s and `typeCast`s to allow void instead of restricting `castPattern`s to not take void. https://github.com/dart-lang/language/issues/2898

**Comment by lrhn** [comment]
We should probably *fix* that, and make cast patterns use `typeNotVoid`.

There is no benefit in allowing `as void`, and we should not allow you to capture or inspect a `void`-typed value at all.

**Comment by modulovalue** [comment]
I have reopened this issue because Lasse has provided an alternative approach that differs from #2898, but one that matches the initial description of this issue.

**Comment by munificent** [comment]
I honestly don't care much one way or the other. If a user writes `as void` and the language semantics are well-defined (which I believe they are), then I don't mind if the language allows it. But if we wanted to forbid it since it's useless, that's OK too.

But, for better or worse, we're not going to change this in the initial release.
